### DISCUSSION TITLE
How to useFetcher in v7?

### DISCUSSION THREAD

**Discussion by Eldevia** [remix-run/remix#10362] [Q&A]
Hi, if i use V7 Framework. Do i need to use browserrouter or is there better way?

**Comment by Eldevia** [comment]
i use fetcher in context to fetch csrf and like login/auth stuff

**Comment by sergiodxa** [comment]
If you use library mode you need to use createBrowserRouter and not BrowserRouter

**Reply by Eldevia** [reply]
i migrated from remix, i wanna use as framework

**Reply by sergiodxa** [reply]
Then it's the same as in Remix

**Reply by Eldevia** [reply]
i get error: useFetcher must be used within a data router.  See https://reactrouter.com/en/main/routers/picking-a-router.

but this error nowhere documented in the v7 docs even the link is invalid!

**Reply by sergiodxa** [reply]
Are you using React Router v7 Framework mode? It's hard to tell without seeing any actual code, but you could create a new React Router app using the CLI and add a fetcher there to see it works, then compare that with your app.

Ensure the component using useFetcher is inside your RR app (inside routes) and not outside, if you have something like:

```ts
import { StrictMode, startTransition } from "react";
import { hydrateRoot } from "react-dom/client";
import { HydratedRouter } from "react-router/dom";

startTransition(() => {
  hydrateRoot(
    document,
    <StrictMode>
      <SomeProvider>
        <HydratedRouter />
      </SomeProvider>
    </StrictMode>,
  );
});
```

The SomeProvider component and any component rendered inside it will be outside your router, and in that case it won't be able to use useFetcher or any other RR feature.
